akka-sample-distributed-workers-scala/src/main/scala/worker/WorkManager.scala

```diff
@@ -19,169 +19,124 @@ import worker.WorkState.WorkerFailed
+  private case class RequestNextWrapper(ask: RequestNext[WorkerCommand]) extends Command
+  final case class WorkIsDone(id: String) extends Command
+  final case class WorkFailed(id: String, t: Throwable) extends Command
+  private final case class TryStartWork() extends Command
 
   def apply(workTimeout: FiniteDuration): Behavior[Command] =
     Behaviors.setup { ctx =>
       Behaviors.withTimers { timers =>
         // FIXME use typed pub sub once https://github.com/akka/akka/issues/26338 is done
+
+        implicit val timeout = Timeout(5.seconds)
         val mediator = DistributedPubSub(ctx.system.toClassic).mediator
+        val producerController = ctx.spawn(WorkPullingProducerController[WorkerCommand]("work-manager", ManagerServiceKey, None), "producer-controller")
+        val requestNextAdapter = ctx.messageAdapter(RequestNextWrapper)
+        producerController ! WorkPullingProducerController.Start(requestNextAdapter)
 
-        // the set of available workers is not event sourced as it depends on the current set of workers
-        var workers = Map[ActorRef[WorkerCommand], WorkerState]()
+        var requestNext = Queue[RequestNext[WorkerCommand]]()
 
-        def notifyWorkers(workState: WorkState): Unit =
+        def tryStartWork(workState: WorkState): Effect[WorkDomainEvent, WorkState] = {
           if (workState.hasWork) {
-            workers.foreach {
-              case (_, WorkerState(ref, Idle)) =>
-                ref ! WorkAvailable
-              case _ => // busy
+            requestNext match {
+              case next +: xs =>
+                val work = workState.nextWork
+                ctx.ask[MessageWithConfirmation[WorkerCommand], Done](next.askNextTo, done => MessageWithConfirmation(DoWork(work), done)) {
+                  case Success(Done) =>
+                    WorkIsDone(work.workId)
+                  case Failure(t) =>
+                    ctx.log.error("Work failed", t)
+                    WorkFailed(work.workId, t)
+                }
+                requestNext = xs
+                Effect.persist(WorkStarted(work.workId))
+              case _ =>
+                Effect.none
             }
+          } else {
+            Effect.none
           }
-
-        def changeWorkerToIdle(worker: ActorRef[WorkerCommand], workId: String): Unit =
-          workers.get(worker) match {
-            case Some(workerState @ WorkerState(_, Busy(`workId`, _))) =>
-              val newWorkerState = workerState.copy(status = Idle)
-              workers += (worker -> newWorkerState)
-            case _ =>
-            // ok, might happen after standby recovery, worker state is not persisted
-          }
-
-        val listingResponseAdapter = ctx.messageAdapter[Receptionist.Listing](UpdatedWorkers)
-        ctx.system.receptionist ! Receptionist.Subscribe(WorkManager.WorkerServiceKey, listingResponseAdapter)
+        }
 
         EventSourcedBehavior[Command, WorkDomainEvent, WorkState](
```

Comment:
This is a nice example where it's good to not use the durableQueue in WorkPullingConsumerController, and instead keep the durable state in the application domain.

An additional experiment would be to see what it would look like to only use the durableQueue in WorkPullingConsumerController. What functionality would be missing in the WorkManager then? The idempotent SubmitWork I guess.

Reply:
Yes perhaps this sample should just use the durable queue instead to keep it simple. I'll try that next week